### PR TITLE
Move onStart() call to choose method.

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**

### CODE REVIEW

**Comment by spencergibb** [comment]
Why the move?

**Comment by OlgaMaciaszek** [comment]
So for instances where the `choose()` method is called alone (for example in Feign load-balanced clients), the suppliers will be called and whoever calls this method does not have to remember to trigger the suppliers, but it's done automatically by the LB.

**Comment by OlgaMaciaszek** [comment]
Also, have renamed some of the request context classes to have `DefaultRequestContext` at the base of the hierarchy so that it's more intuitive for people that will be creating their `LoadBalancerLifecycle` classes.

**Comment by spencergibb** [comment]
I think this complicates the idea of `LoadBalancerLifecycle`. We want folks to **not** use `choose()` directly. It should be documented that if they do, they are responsible for lifecycle method calls. Since we own the feign impl, we can do that. We should add tests there because it seems like this is a response to forgetting to make those call.

TLDR don't merge this and add some docs and javadocs.

**Comment by OlgaMaciaszek** [comment]
Ok, you've convinced me about not moving it. Will close this.
